packages/core/test/unit/Site.test.js

```diff
@@ -676,3 +682,212 @@ siteJsonPageExclusionTestCases.forEach((testCase) => {
+  const json = {
+    ...PAGE_NJK,
+    'site.json': SITE_JSON_DEFAULT,
+    '_markbind/versions.json': VERSIONS_DEFAULT,
+  };
+  fs.vol.fromJSON(json, '');
+
+  const site = new Site('./', '_site');
+  await site.readSiteConfig();
+
+  expect.assertions(1);
+  try {
+    await site.writeVersionsFile('v2', 'custom/archive/path');
+  } catch (e) {
+    expect(e).toEqual(new Error('The version name is the same as a previously archived version, but the'
+      + ' archive path is not. This is likely to be an error as the previous version will'
+      + ' no longer be tracked and managed. Please choose a different name or manually'
+      + ' change the clashing name in the versions.json file to a different name'));
+  }
+});
+
+test('Site correctly updates the versions in the versions file for an overwritten version', async () => {
+  const json = {
+    ...PAGE_NJK,
+    'site.json': SITE_JSON_DEFAULT,
+    '_markbind/versions.json': VERSIONS_DEFAULT,
+  };
+  fs.vol.fromJSON(json, '');
+
+  const site = new Site('./', '_site');
+  await site.readSiteConfig();
+  const someVersionsData = await site.writeVersionsFile(
+    'testOverwritingVersion', 'version/testOverwritingVersion');
+  const newVersionData = {
+    versionName: 'testOverwritingVersion',
+    buildVer: MARKBIND_VERSION,
+    archivePath: 'version/testOverwritingVersion',
+    baseUrl: JSON.parse(SITE_JSON_DEFAULT).baseUrl,
+  };
+
+  const expectedVersionsData = JSON.parse(VERSIONS_DEFAULT);
+  expectedVersionsData.versions[2] = newVersionData; // because constant is hardcoded
+
+  expect(someVersionsData).toEqual(expectedVersionsData);
+});
+
+const copyingArchivedSiteTestCases = [
+  {
+    name: 'No versions copied when none are specified in site.json',
+    versionsToGenerate: false,
+    versions: [],
+    expected:
+      {
+        v1: false,
+        v2: false,
+        differentBaseUrl: false,
+      },
+  },
+  {
+    name: 'Versions specified in site.json are copied over',
+    versionsToGenerate: false,
+    versions: ['v2'],
+    expected:
+      {
+        v1: false,
+        v2: true,
+        differentBaseUrl: false,
+      },
+  },
+  {
+    name: 'The different versions specified in flag override those in site.json',
+    versionsToGenerate: ['v1'],
+    versions: ['v2'],
+    expected:
+      {
+        v1: true,
+        v2: false,
+        differentBaseUrl: false,
+      },
+  },
+  {
+    name: 'Versions flag overrides versions specified in site.json',
+    versionsToGenerate: true,
+    versions: ['v2'],
+    expected:
+      {
+        v1: false,
+        v2: false,
+        differentBaseUrl: false,
+      },
+  },
+  {
+    name: 'Even when specified, the version with a differentBaseUrl is not deployed',
+    versionsToGenerate: false,
+    versions: ['differentBaseUrl'],
+    expected:
+      {
+        v1: false,
+        v2: false,
+        differentBaseUrl: false,
+      },
+  },
+];
+
+// TODO:
```

Comment:
> We could use a functional test with some "special" procedures for this (much like the special procedures for testing markbind convert vs other sites)

Thanks for this. I was actually just expecting functional tests `test/e2e/functional` which are easier to maintain.
For example,

```js
// testSites.js

const testVersioningSites = [
  'test_site_versioning/1',
];

// test.js
testVersioningSites .forEach((siteName) => {
  console.log(`Running ${siteName} versioning tests`);
  const nonMarkBindSitePath = path.join(siteName, 'non_markbind_site');
  try {
    execSync(`node ../../index.js version versionName`, execOptions);
    compare(siteName, 'expected1', '_site');

    // The downside specific to MarkBind's functional tests is that we need the expected directory (which if there are too many of might begin to slow git a little)
   // But you don't have to create one site for every behaviour you want to test,
   // e.g. get creative with something like this

    execSync(`node ../../index.js version versionName --versions`, execOptions);
    expect versions dir to be empty

    ... more ...
    
  } catch (err) {
    printFailedMessage(err, siteName);
    cleanupConvert(path.resolve(__dirname, siteName));
    process.exit(1);
  }
  cleanupConvert(path.resolve(__dirname, siteName));
});
```

In general, note that we aim to test public interfaces instead of private ones. In the case of markbind (a cli app), especially about `Site.js`'s functionalities which integrate multiple things, that's predominantly e2e behaviour.

You may have noticed the tests are quite brittle (e.g. what if we reordered the execution of `readSiteConfig` / `writeVersionsFile`), these are whimsical implementation details that once changed would require changing the tests too.

**A lot of our existing unit tests** frankly don't adhere to this well too, and definitely need improving.

Though for clarification -- this is not me making a case for e2e tests, just "which is easier to write and maintain" _in this case_ (and still tests what we want to test), which varies from project to project. For example, deeper implementation details that are sufficiently complex and well encapsulated, e.g. a hypothetical pure function like `replaceInvalidCharacters`, can and probably should be unit-tested, if we expect them to work a certain way.

I'm ok with keeping the tests here since you've already done quite a bit of work here, but **if you have the time**, I think we can adapt these to functional tests. But we can definitely work on porting the tested behaviours otherwise in the future :)

Reply:
Whoops... Maybe for a future PR 😓  I don't think I'll have the time in the coming months.